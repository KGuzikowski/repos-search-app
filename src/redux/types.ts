import { ThunkAction } from 'redux-thunk'
import { Action } from 'redux'

export type AppThunk<State, ReturnType = void> = ThunkAction<ReturnType, State, unknown, Action<string>>