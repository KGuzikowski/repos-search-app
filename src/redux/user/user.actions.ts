import { userActionTypes, userStateType, loginStartType, loginSuccessType, loginFailureType, userType, logoutType } from './user.types'
import { ThunkAction } from 'redux-thunk'
import { Action } from 'redux'

export type AppThunk<State, ReturnType = void> = ThunkAction<ReturnType, State, unknown, Action<string>>

const loginStart = (): loginStartType => ({
    type: userActionTypes.LOGIN_START
})

const loginSuccess = (user: userType, accessToken: string): loginSuccessType => ({
    type: userActionTypes.LOGIN_SUCCESS,
    payload: {
        user,
        accessToken
    }
})

const loginFailure = (errorMessage: string): loginFailureType => ({
    type: userActionTypes.LOGIN_FAILURE,
    payload: errorMessage
})

export const loginUserStartAsync = (code: string): ThunkAction<void, userStateType, unknown, Action<string>> => {
    return async dispatch => {
        try {
            if(!code) throw Error

            dispatch(loginStart())

            const url = `https://cors-anywhere.herokuapp.com/https://github.com/login/oauth/access_token?client_id=${process.env.REACT_APP_CLIENT_ID}&client_secret=${process.env.REACT_APP_CLIENT_SECRET}&code=${code}`
            const access = await fetch(url, {
                method: 'POST',
                headers: {
                    "X-Requested-With": "XMLHttpRequest",
                    'Content-Type': 'application/json',
                    'Accept': 'application/json'
                }
            })

            const { access_token } = await access.json()

            if(!access_token) throw Error

            const urlUser = `https://api.github.com/user`
            const dataUser = await fetch(urlUser, {
                method: 'GET',
                headers: {
                    Authorization: `Bearer ${access_token}`,
                    'Content-Type': 'application/json',
                    'Accept': 'application/json'
                }
            })

            const user = await dataUser.json()

            if(user.error) throw Error

            dispatch(loginSuccess({ login: user.login, id: user.id }, access_token))

        } catch(error) {
            dispatch(loginFailure('Sorry, cannot log in! Please try again.'))
        }
    }
}

export const logout = (): logoutType => ({
    type: userActionTypes.LOGOUT
})