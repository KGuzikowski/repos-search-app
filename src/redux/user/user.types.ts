export type userType = {
    login: string,
    id: number
}

export interface userStateType {
    isFetching: boolean,
    accessToken: string,
    errorMessage: string,
    user: userType | null
}

export const userActionTypes = {
    LOGIN_START: 'LOGIN_START',
    LOGIN_SUCCESS: 'LOGIN_SUCCESS',
    LOGIN_FAILURE: 'LOGIN_FAILURE',
    LOGOUT: 'LOGOUT'
}

export interface loginStartType {
    type: typeof userActionTypes.LOGIN_START
}

export interface loginSuccessType {
    type: typeof userActionTypes.LOGIN_SUCCESS,
    payload: any
}

export interface loginFailureType {
    type: typeof userActionTypes.LOGIN_FAILURE,
    payload: string
}

export interface logoutType {
    type: typeof userActionTypes.LOGOUT
}
