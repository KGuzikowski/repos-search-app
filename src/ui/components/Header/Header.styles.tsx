import styled from 'styled-components'
import { primaryColor } from '../../style-variables'

export const NavBar = styled.nav`
    display: flex;
    justify-content: space-between;
    align-items: center;

    @media only screen and (max-width: 400px) {
        flex-direction: column;
    }
`
export const HeaderTag = styled.header`
    padding: 20px 100px;
    background: ${primaryColor};

    @media only screen and (max-width: 700px) {
        padding: 10px 50px;
    }
`