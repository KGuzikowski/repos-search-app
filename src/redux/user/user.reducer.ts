import { userStateType, userActionTypes } from './user.types'

export const INITIAL_STATE: userStateType = {
    isFetching: false,
    accessToken: '',
    errorMessage: '',
    user: null
}

const userReducer = (state = INITIAL_STATE, action: any): userStateType => {
    switch (action.type) {
        case userActionTypes.LOGIN_START:
            return {
                ...state,
                isFetching: true,
                errorMessage: ''
            }
        case userActionTypes.LOGIN_SUCCESS:
            return {
                ...state,
                isFetching: false,
                errorMessage: '',
                accessToken: action.payload.accessToken,
                user: action.payload.user
            }
        case userActionTypes.LOGIN_FAILURE:
            return {
                ...state,
                isFetching: false,
                errorMessage: action.payload
            }
        case userActionTypes.LOGOUT:
            return {
                ...state,
                accessToken: '',
                user: null
            }
        default: return state
    }
}

export default userReducer