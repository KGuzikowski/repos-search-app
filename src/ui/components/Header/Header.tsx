import React from 'react'
import { useSelector, useDispatch } from 'react-redux'
import { logout } from '../../../redux/user/user.actions'
import { RootState } from '../../../redux/rootReducer'
import { HeaderTag, NavBar } from './Header.styles'
import Logo from '../Logo/Logo'
import Button from '../Button/Button'

const Header = () => {
    const user = useSelector((state: RootState) => state.user)
    const dispatch = useDispatch()

    const handleLogout = (e: React.MouseEvent<HTMLAnchorElement>) => {
        e.preventDefault();
        dispatch(logout())
    }
    
    const href = `https://github.com/login/oauth/authorize?scope=read:user&client_id=${process.env.REACT_APP_CLIENT_ID}`
    return (
        <HeaderTag>
            <NavBar>
                <Logo />
                { !user.accessToken
                    ? <Button href={href}>Log in with GitHub</Button>
                    : <Button onClick={e => handleLogout(e)}>Log out</Button>
                }
            </NavBar>
        </HeaderTag>
    )
}

export default Header