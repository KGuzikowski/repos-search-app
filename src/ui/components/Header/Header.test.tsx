import React from 'react'
import { create } from 'react-test-renderer'
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'
import Header from './Header'
import { userStateType } from '../../../redux/user/user.types'

it('renders Header component correctly', () => {
    const INITIAL_STATE: userStateType = {
        isFetching: false,
        accessToken: 'a22',
        errorMessage: '',
        user: {
            id: 1,
            login: 'Karol'
        }
    }

    const mockStore = configureStore()
    const store = mockStore(INITIAL_STATE)

    const component = create(<Provider store={store}><Header /></Provider>)
    expect(component.toJSON()).toMatchSnapshot()
})