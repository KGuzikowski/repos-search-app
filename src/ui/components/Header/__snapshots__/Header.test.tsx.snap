// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Header component correctly 1`] = `
<header
  className="sc-AxirZ eZIgto"
>
  <nav
    className="sc-AxjAm beKDTQ"
  >
    <div
      className="sc-AxiKw gyEHAk"
    >
      ForCallstack
    </div>
    <a
      className="sc-AxhCb HHzFB"
      href="https://github.com/login/oauth/authorize?scope=read:user&client_id=5065e74296f19a16677e"
    >
      Log in with GitHub
    </a>
  </nav>
</header>
`;
